docs/development/pigs/pig-013.rst

```diff
@@ -0,0 +1,451 @@
+support this for the coming years.
+
+Required dependencies
+=====================
+
+We propose to update the Gammapy required dependencies as shown in the following
+table (the release dates for the packages are shown in parentheses are were
+obtained from https://pypi.org/). A definition of required dependency is given
+above in the introduction, notes on each required dependency are given below
+after the table.
+
+===============  ================  ================
+Dependency       Gammapy 0.12      Gammapy 0.14
+===============  ================  ================
+Python           3.5   (Sep 2015)  3.6   (Dec 2016)
+Numpy            1.10  (May 2016)  1.16  (Jan 2019)
+Scipy            0.15  (Jan 2015)  1.2   (Dec 2018)
+Astropy          2.0   (Jul 2017)  3.2   (Jun 2019)
+regions          0.3   (Sep 2018)  0.5   (Jul 2019)
+pyyaml           unclear           3.13  (Jul 2018)
+click            unclear           7.0   (Sep 2018)
+===============  ================  ================
+
+We already mentioned the possibility to drop Python 3.5 support in
+:ref:`pig-003`. One reason is that Anaconda and conda-forge (our main
+distribution channel, used in our testing continuous integration setup) only
+contains Python 2.7, 3.6 and 3.7 at this point (and 3.8 added in fall 2018),
+i.e. testing on Python 3.5 is already extra effort. Also, Python 3.6 allows us
+to use f-strings and just dict instead of a mix of dict and OrderedDict like we
+have now, it's nice for developers. E.g. Sunpy or ctapipe already require Python
+3.6 or later.
+
+This update is to very recent versions for dependencies. We could keep support
+for older versions, e.g. Numpy 1.13 (Jun 2017), Scipy 1.0  (Oct 2017) and
+Astropy 3.0 (Feb 2018) or even older. But it wasn't clear to us if doing so
+would be of use to anyone, so we defaulted to this proposal.
+
+For Astropy, we could keep support for version 2.0, but there are many small new
+features, fixes and performance improvements in 3.0, 3.1 and 3.2 (see `Astropy
+whatsnew`_) that will make our life easier, for example much improved
+Quantity-support for functions in Numpy and Scipy, support for the FITS time
+standard and better support for Time columns in Table, the new
+``astropy.uncertainty`` package which we could use (see :ref:`pig-14`_), the
+``SkyCoord.directional_offset_by`` method for field of view coordinates, common
+WCS API, the new ``astropy.timeseries``, ASDF.
+
+A major motivation to update to very recent versions is that the ``regions``
+package is still under development (see :ref:`pig-010`_). In Gammapy 0.13 we
+require regions 0.4, and we plan to make a regions 0.5 release with further
+features and fixes in July, and to require that for Gammapy 0.14.
+
+The ``pyyaml`` package is the most-used YAML package for Python, used e.g. also
+in Astropy core for ECSV or ASDF I/O. The ``click`` package was chosen to
+implement the ``gammapy`` command line interface. For ``pyyaml`` and ``click``
```

Comment:
I would strongly recommend against pyyaml. It uses an outdated spec (1.1) of yaml and will not be updated to support a newer one. 

Conda uses ruamel_yaml, which supports yaml 1.2.

I think this especially important for astropy/gammapy as only yaml 1.2 supports scientific notation floats like `1.5e3`

Reply:
@MaxNoe -  could you also raise this as an issue on the astropy issue tracker?

Reply:
@maxNoe - Thank you for the suggestion! That is exactly the kind of discussion and feedback I had hoped to trigger via this PIG.

I'm not sure if `pyyaml` or `ruamel_yaml` is the better choice, still suggest to stick with `pyyaml` for now.

Yes, `ruamel_yaml` has some nice things, and it's weird that pyyaml doesn't parse `1.5e3` as a float, but requires one to write `1.5e+3` instead. The round-tripping feature including comments is for me the biggest improvement over `pyyaml`, that's really nice if we want to read and write back modified YAML files ever in a way that is as unintrusive as possible (not clear if we need that - maybe for config files).

On the other hand, with `ruamel_yaml` I have concerns with stability and support. E.g. https://pypi.org/project/ruamel.yaml/ says **For production systems you should pin the version being used with ``ruamel.yaml<=0.15** and it's not really clear to me if / what they plan to change in v0.16 or v1.0 for `ruamel_yaml`, and if that ever will come or not, i.e. if and when things would break in Gammapy. It looks like it's mostly one guy developing `ruamel_yaml` on Bitbucket and he has ~ 100 packages with `ruamel` in the name, and I'm not sure what that is?

Astropy and ASDF use `pyyaml`, in Astropy for ECSV and other things. It seems stable and well-supported, there's conda binaries and pip wheels and I never had an install issue over the past years.

So if we adopt `ruamel_yaml` in Gammapy, we effectively depend on both, in addition to `pyyaml`, kind of doubling the chance of trouble for us in the future (e.g. API changes, or packaging screw-ups, ...). For that reason I would suggest we stick with `pyyaml` for now, and only mention the situation and `ruamel_yaml` in the alternatives section, as something we might change to later if Astropy and the larger Python community adopts it over `pyyaml` (which might or might not happen).

Reply:
0.16 is already out, the changed the API from 

```
from ruamel.yaml import safe_load # or load
config = safe_load(open('config.yaml'))
```

to a classed based loader, that takes configuration like this:

```
from ruamel.yaml import YAML

yaml = YAML(typ='safe')
config = yaml.load(open('config.yaml')
```

As mentionend, conda itself already depends on ruamel yaml, so it's basically guarenteed to be available with conda and no problem to get it with pip.

Reply:
The changes announced are the removal of the old api, which is still supported but deprecated in 0.15.

Reply:
I've split pyyaml and click into their own paragraphs, it wasn't good to mix them into one.

@registerrier - for click I've said it will be re-considered in the high-level PIG. That one will probably take longer, shouldn't hold up this one. We can implement this and then still change our mind in the coming months, that's fine. The key proposal here is to drop support for the old versions, and because that might be controversial I went for a PIG.

@maxnoe - for YAML I have a preference to stick with `pyyaml`, sorry.
I've added a longish section trying to summarise the pros and cons of `pyyaml` and `ruamel_yaml` and stated why we prefer to stick with `pyyaml` for now.
See 0080a4bc6 .
This PIG will go to official review phase soon, and we might re-discuss and re-consider this, but for now I'm marking this discussion as resolved.